Freertos working commands

Known4225 opened the pull request:
This temporary branch implements the following features in freertos:
- uart communication
- command handler
- blink app

Known4225 committed:
temp branch to reduce file changes

Known4225 committed:
checking if git is smart enough to figure this out

Known4225 committed:
some files removed from appcpu1 so they show up as renames in the shared folder

Known4225 commented:
@npetersen2 ready for review

@codecubepi please review code location under the `shared/` folder

npetersen2 commented on sdk/FreeRTOS-Kernel:
It appears the `.gitmodules` has to been changed, so why this change for the submodule here?

We should be trying to keep to [named version releases](https://github.com/FreeRTOS/FreeRTOS-Kernel/releases) from FreeRTOS---which version number does this change correspond to?

npetersen2 commented on sdk/shared/sys/serial.h:
```diff
@@ -1,13 +1,11 @@
-
 #define SERIAL_UPDATES_PER_SEC (10000)
-#define SERIAL_INTERVAL_USEC   (USEC_IN_SEC / SERIAL_UPDATES_PER_SEC)
+#define SERIAL_INTERVAL_TICKS  (pdMS_TO_TICKS(1000.0 / SERIAL_UPDATES_PER_SEC))
```
Include header for `pdMS_TO_TICKS` macro

npetersen2 commented on sdk/shared/sys/serial.c:
```diff
@@ -18,38 +21,42 @@ static int print_idx = -1;
-    scheduler_tcb_register(&tcb);
+	// Create serial task
+	xTaskCreate(serial_main, (const char *) "uartSerial", configMINIMAL_STACK_SIZE,
+				NULL, tskIDLE_PRIORITY, &xSerialTaskHandle);
```
Let's make a new header file that has all the task priorities in one place so we can reason about how the system operates easily.

Let's call it `sys/task_priority.h` and define priority there, then include and reference these defines in this file from each place that tasks are created.

npetersen2 replied:
Note: this comment applies to all the FreeRTOS tasks you have added, not just this one! There should be numerous priorities defines in this new file.

npetersen2 commented on sdk/shared/sys/intr.h:
```diff
@@ -1,51 +1,51 @@
+#include <stdint.h>
+#include <stdio.h>
+
+// test
```
Remove this line

npetersen2 commented on sdk/shared/sys/intr.h:
```diff
@@ -1,51 +1,51 @@
+#define INTC_1TO0_SEND_INTERRUPT_ID 2U
+#define INTC_0TO1_RCVE_INTERRUPT_ID 3U
+
+// Interrupt Controller Instance
+//   Defined here to be accessable in sys/icc.c
+XScuGic InterruptController;
```
Scary scary! We should not be defining objects in header files.

Who "owns" this object? Where does this object live?

This code probably already was there before this PR... but I see you moved `static`.

Either way, we need to move this to a single `.c` file, then expose a `extern` declaration in the header

codecubepi replied:
This is probably my bad...

npetersen2 commented on sdk/shared/sys/intr.c:
I think there are no changes in this file

npetersen2 commented on sdk/shared/sys/icc.h:
I think there are no changes in this file

npetersen2 commented on sdk/shared/sys/icc.c:
I think there are no changes in this file

npetersen2 commented on sdk/shared/sys/commands.h:
```diff
@@ -1,10 +1,8 @@
-
 #define COMMANDS_UPDATES_PER_SEC (10000)
-#define COMMANDS_INTERVAL_USEC   (USEC_IN_SEC / COMMANDS_UPDATES_PER_SEC)
+#define COMMANDS_INTERVAL_TICKS  (pdMS_TO_TICKS(1000.0 / COMMANDS_UPDATES_PER_SEC))
```
Include file which exposes `pdMS_TO_TICKS`

npetersen2 commented on sdk/shared/sys/commands.c:
I reviewed this file, it looks like, in general, you didn't change anything except for commenting out code related to ethernet and icc.

Overall, seems reasonable for now. But, we'll need to comb through this again in the future and make sure things are cleaned up

npetersen2 commented on sdk/shared/platform_config.h:
No changes here

npetersen2 commented on sdk/shared/drv/uart.c:
```diff
@@ -1,4 +1,4 @@
-#include "drv/uart.h"
```
Why this change? Let's keep the full path so it is clear where the header file lives and its "type"

npetersen2 commented on sdk/shared/README.md:
No changes here

codecubepi replied:
Line endings

npetersen2 commented on sdk/freertos_app_cpu1/.project:
```diff
@@ -1,38 +1,43 @@
+		<nature>org.eclipse.cdt.managedbuilder.core.ScannerConfigNature</nature>
+	</natures>
+	<linkedResources>
+		<link>
+			<name>FreeRTOS-Kernel</name>
+			<type>2</type>
+			<location>C:/Users/srich008/Information/v2-staging/sdk/FreeRTOS-Kernel</location>
+		</link>
+		<link>
+			<name>MemMang</name>
+			<type>2</type>
+			<location>C:/Users/srich008/Information/v2-staging/sdk/FreeRTOS-Kernel/portable/MemMang</location>
+		</link>
+		<link>
+			<name>shared</name>
+			<type>2</type>
+			<location>C:/Users/srich008/Information/v2-staging/sdk/shared</location>
+		</link>
```
We'll need to clean this up to not have absolute paths on your computer.

We can change these with relative project paths if you are careful in the SDK wizard...

For now, let's ignore this and keep moving forward.

codecubepi replied:
Yes, I added a similar comment on this file in the `cpu0` app.

npetersen2 commented on sdk/freertos_app_cpu0/src/usr/user_config.h:
Interesting, why this empty file?

I suppose future modules you will port over will need this?

codecubepi replied:
It looks like this is replacing the old `user_config.h` file? Perhaps this should be a `shared/` resource going forward?

npetersen2 commented on sdk/freertos_app_cpu0/src/usr/game/task_game.h:
```diff
@@ -0,0 +1,48 @@
+// Note that it can be overridden via usr/user_config.h
+#define TASK_GAME_UPDATES_PER_SEC (5)
+
+// Microseconds interval between when task is called
+//
+// This is what scheduler actually uses to run task,
+// but is generated via define above
+#define TASK_GAME_INTERVAL_USEC (USEC_IN_SEC / TASK_GAME_UPDATES_PER_SEC)
```
I noticed your game task has a different way of defining its `interval_usec` stuff versus the other system tasks.

Eventually, we should be consistent on the "recommended" way of creating tasks.

If we want your game to be kept in the codebase for the full `v2` release, we'll need to update this. Otherwise, if the game will get removed, it does not matter

codecubepi replied:
Yeah, do we need the "game" in the codebase? This could become confusing clutter in the future if it's not a required resource.

npetersen2 commented on sdk/freertos_app_cpu0/src/usr/game/cmd/cmd_game.c:
Interesting commands here :)

npetersen2 commented on sdk/freertos_app_cpu0/src/main.c:
```diff
@@ -211,150 +135,7 @@ int main(void)
-    } else {
-        message_status = 2;
-        xil_printf("CPU0 - FreeRTOS Hello World Example FAILED\r\n");
+    	xil_printf("IF YOU'RE READING THIS THEN A TERRIBLE ERROR HAS OCCURRED!\n");
```
Ha nice. :)

I am happy to keep this, but let's give the user a little more help here. Revise to say something about memory

codecubepi replied:
I need the deleted code to be kept as well, see above comment on deleted code. I think you can probably add a task for your own testing without deleting my tasks.

npetersen2 commented on sdk/freertos_app_cpu0/src/main.c:
```diff
@@ -91,50 +87,12 @@ extern void vPortInstallFreeRTOSVectorTable(void);
     // S=b1 TEX=b100 AP=b11, Domain=b1111, C=b0, B=b0
     Xil_SetTlbAttributes(0xFFFF0000, 0x14de2);
 
-#if 1
-    // CPU0 ONLY:
-    // This code is required to start CPU1 from CPU0 during boot.
-    //
-    // This only applies when booting from flash via the FSBL.
-    // During development with JTAG loading, these low-level
-    // calls in this #if block are not needed! However, we'll
-    // keep them here since it doesn't affect performance...
-
-    // Write starting base address for CPU1 PC.
-    // It will look for this address upon waking up
-    static const uintptr_t CPU1_START_ADDR = 0xFFFFFFF0;
-    static const uint32_t CPU1_BASE_ADDR = 0x20080000;
-    Xil_Out32(CPU1_START_ADDR, CPU1_BASE_ADDR);
-
-    // Waits until write has finished
-    // DMB = Data Memory Barrier
-    dmb();
-
-    // Wake up CPU1 by sending the SEV command
-    // SEV = Set Event, which causes CPU1 to wake up and jump to CPU1_BASE_ADDR
-    __asm__("sev");
-#endif
-
```
Let's keep this code

codecubepi replied:
Yes please keep... if this is deleted, we'll get nasty merge conflicts when I merge the dual-core stuff.

npetersen2 reviewed (changes requested):
Thanks @Known4225 , I think this is great! I reviewed each file that changed and left some comments as appropriate.

Overall, there are several small things I'd like you to do, but in general, this is looking pretty close to merging from my perspective.

Let's see what @codecubepi has to say in his review

codecubepi commented on sdk/freertos_app_cpu0/src/usr/user_apps.h:
@npetersen2 Perhaps this should be a `shared/` resource going forward?

codecubepi commented on sdk/freertos_app_cpu0/src/usr/user_apps.c:
Perhaps this should be a `shared/` resource?

codecubepi commented on sdk/freertos_app_cpu0/.project:
Hmmm... we really need to determine a way to ditch these absolute paths :(

Probably fine for now though.

codecubepi commented on sdk/freertos_app_cpu0/.cproject:
This whole file is replaced and regenerated when a new user creates the project... perhaps we can delete it?

codecubepi commented on sdk/freertos_app_cpu0/src/FreeRTOSConfig.h:
```diff
@@ -1,247 +1,251 @@
+#define configNET_MASK2 255
+#define configNET_MASK3 0
+
+/* MS to tick macros */
+#define pdMS_TO_TICKS(xTimeInMs)    ((TickType_t) (((xTimeInMs) * (uint64_t) configTICK_RATE_HZ) / (uint64_t) 1000U))
+#define pdTICKS_TO_MS(xTimeInTicks)    (((xTimeInTicks) * (uint64_t) 1000U) / (double) configTICK_RATE_HZ)
+
```
@Known4225 

Can you confirm that these added lines are the only changes to this file? The diff shows that the whole file was changed, why? It looks like this is the only addition.

codecubepi commented on sdk/freertos_app_cpu0/src/Xilinx.spec:
Line ending change?

codecubepi commented on sdk/freertos_app_cpu0/src/main.c:
```diff
@@ -410,3 +191,39 @@ void vApplicationGetTimerTaskMemory(StaticTask_t **ppxTimerTaskTCBBuffer,
     *pulTimerTaskStackSize = configTIMER_TASK_STACK_DEPTH;
 }
+
+/*******************************************************************************/
+/*              S E T U P   I N T E R R U P T   S Y S T E M                    */
+/*******************************************************************************/
+int SetupInterruptSystem(XScuGic *IntcInstancePtr) {
+	int Result;
+
+	XScuGic_Config *IntcConfig;
+
+	// Initialize the interrupt controller driver so that it is ready to use.
+
+	IntcConfig = XScuGic_LookupConfig(INTC_DEVICE_ID);//ELS -- this is misleading! there is only one XScuGic device (0)
+	if (NULL == IntcConfig) {
+		return XST_FAILURE;
+	}
```
What is the purpose of adding this here? This is done in `intr_init()` in `shared/sys/intr.c`

codecubepi commented on sdk/freertos_app_cpu1/.cproject:
See my comment on `freertos_app_cpu0/.cproject`

codecubepi commented on sdk/freertos_app_cpu1/src/FreeRTOSConfig.h:
Again, what was changed in this file? Line endings? 

@Known4225 please be conscious when you commit files of exactly what is changing. I recommend using the Diff View under 'Source Control' in VS Code... 

![image](https://github.com/user-attachments/assets/738bfaf1-fb73-4e25-be9d-5a8fbc218bb9)

If the only thing changing is line endings, please don't commit the change :)

codecubepi commented on sdk/freertos_app_cpu1/src/Xilinx.spec:
See other comments on line endings

codecubepi commented on sdk/shared/FreeRTOS_tick_config.c:
Line endings?

codecubepi reviewed (changes requested):
@Known4225 

Good work so far. However it looks like many of your files have either changed minimally or not at all, but are marked as changed, probably due to line endings. If you could revert these EOL changes it will minimize the number of files changed, make review much easier, and Git will not have to store/track meaningless changes.

Known4225 committed:
removed game app

Known4225 committed:
readded queue and blinky tasks to main.c

Known4225 committed:
converted LF to CRLF in shared/sys files

Known4225 committed:
okay that didn't work

Known4225 committed:
linux permissions

Known4225 committed:
line endings on cpu0

Known4225 committed:
line endings on cpu1 and shared

Known4225 committed:
permissions on cpu1

Known4225 committed:
linux permissions on shared

Known4225 committed:
permissions on blink

Known4225 committed:
platform config

Known4225 committed:
include projdefs in commands.h

Known4225 committed:
removed vscode folder

Known4225 committed:
runtime statistics implemented, ran the code on AMDC and the tasks clobber each other and the lights look funny, but this is expected behaviour

Known4225 commented:
Changes:
- created task_priority.h file with task priorities for serial, cmds, and the blink and vsi tasks.
- deleted game app
- added user_config content
- removed all line endings and permission issues (git can't seem to figure out that task_blink.c, task_vsi.c, task_vsi.h, and commands.c have been moved AND changed. So it lists them as deleted and re-added with different contents)
- slight changes to FreeRTOSConfig.h (enabling run time statistics and providing a timer to use)
- Added timer_stats for run time statistics timer
- fixed intr.h header definition (now an extern declaration and definition in the source file)
- included "projdefs.h" in commands.h to provide path to pdMS_TO_TICKS macro
